adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java

```diff
@@ -613,96 +610,86 @@ private String readCode(Reader reader) throws IOException {
 
         private String code;
-
         private String error;
-
         private String errorDescription;
+        private String state;
+        private Undertow server;
 
-        private IOException errorException;
+        private GracefulShutdownHandler gracefulShutdownHandler;
 
-        private String state;
+        public void start() {
+            PathHandler pathHandler = Handlers.path().addExactPath("/", this);
+            AllowedMethodsHandler allowedMethodsHandler = new AllowedMethodsHandler(pathHandler, Methods.GET);
+            gracefulShutdownHandler = Handlers.gracefulShutdown(allowedMethodsHandler);
+
+            server = Undertow.builder()
+                    .addHttpListener(0, "localhost")
+                    .setHandler(gracefulShutdownHandler).build();
 
-        private Socket socket;
+            server.start();
+        }
 
-        private HttpResponseWriter writer;
+        public void stop() {
+            server.stop();
+        }
 
-        public CallbackListener(HttpResponseWriter writer) throws IOException {
-            this.writer = writer;
-            server = new ServerSocket(0);
+        public int getLocalPort() {
+            return ((InetSocketAddress) server.getListenerInfo().get(0).getAddress()).getPort();
+        }
+
+        public void await() throws InterruptedException {
+            shutdownSignal.await();
         }
 
         @Override
-        public void run() {
-            try {
-                socket = server.accept();
-
-                BufferedReader br = new BufferedReader(new InputStreamReader(socket.getInputStream()));
-                String request = br.readLine();
-
-                String url = request.split(" ")[1];
-                if (url.indexOf('?') >= 0) {
-                    url = url.split("\\?")[1];
-                    String[] params = url.split("&");
-
-                    for (String param : params) {
-                        String[] p = param.split("=");
-                        if (p[0].equals(OAuth2Constants.CODE)) {
-                            code = p[1];
-                        } else if (p[0].equals(OAuth2Constants.ERROR)) {
-                            error = p[1];
-                        } else if (p[0].equals("error-description")) {
-                            errorDescription = p[1];
-                        } else if (p[0].equals(OAuth2Constants.STATE)) {
-                            state = p[1];
-                        }
-                    }
-                }
+        public void handleRequest(HttpServerExchange exchange) throws Exception {
+            gracefulShutdownHandler.shutdown();
 
-                OutputStreamWriter out = new OutputStreamWriter(socket.getOutputStream());
-                PrintWriter pw = new PrintWriter(out);
-                if (writer != null) {
-                    System.err.println("Using a writer is deprecated.  Please remove its usage.  This is now handled by endpoint on server");
```

Comment:
We should probably advertise in our change notes the removal of `HttpResponseWriter` from the API?